File: Saving/Saving.cpp
#include "Saving.h"
#include "../Organism/Organism.h"
#include "../World/World.h"
#include "../Human/Human.h"
#include "../Sheep/Sheep.h"
#include "../Antelope/Antelope.h"
#include "../SosnowskyHogweed/SosnowskyHogweed.h"
#include "../Guarana/Guarana.h"
#include "../Grass/Grass.h"
#include "../Wolf/Wolf.h"
#include "../Fox/Fox.h"
#include "../Turtle/Turtle.h"
#include "../NightshadeBerries/NightshadeBerries.h"
#include "../Milkweed/Milkweed.h"
#include <fstream>

// File format:
// 1. World id
// 2. Day number
// 3. World width
// 4. World height
// 5. Board start x
// 6. Board start y
// 7. Human ability duration
// 8. Human ability cooldown
// 9. Organisms count
// 10. Organisms:
//    - Type
//    - Id
//    - Strength
//    - Initiative
//    - Position x
//    - Position y
//    - Parent 1 id
//    - Parent 2 id
//    - Age
//    - Previous x
//    - Previous y
//    - Born
//    - Alive

Saving::Saving(World* world) : world(world) {}
World* Saving::getWorld() const { return this->world; }

void Saving::save(const std::string& fileName) {
  std::ofstream file(fileName + ".txt");

  World* w = this->getWorld();

  file << w->getId() << std::endl;
  file << w->getDay()-1 << std::endl;
  file << w->getWidth() << std::endl;
  file << w->getHeight() << std::endl;
  file << w->getStartX() << std::endl;
  file << w->getStartY() << std::endl;

  file << w->getHuman()->getAbilityDuration() << std::endl;
  file << w->getHuman()->getAbilityCooldown() << std::endl;

  file << w->getOrganisms().size() << std::endl;

  for (Organism* organism : w->getOrganisms()) {
    file << organism->getType() << " ";
    file << organism->getId() << " ";
    file << organism->getStrength() << " ";
    file << organism->getInitiative() << " ";
    file << organism->getPosition().getX() << " ";
    file << organism->getPosition().getY() << " ";
    file << organism->getParent1Id() << " ";
    file << organism->getParent2Id() << " ";
    file << organism->getAge() << " ";
    file << organism->getPreviousX() << " ";
    file << organism->getPreviousY() << " ";
    file << organism->isBorn() << " ";
    file << organism->isAlive() << std::endl;
  }

  file.close();
}

World Saving::load(const std::string &fileName) {
  std::ifstream file(fileName + ".txt");

  std::string wId;
  int day;
  int width;
  int height;
  int startX;
  int startY;
  int humanAbilityDuration;
  int humanAbilityCooldown;
  int organismsCount;
  std::list<Organism*> organisms;

  file >> wId;
  file >> day;
  file >> width;
  file >> height;
  file >> startX;
  file >> startY;
  file >> humanAbilityDuration;
  file >> humanAbilityCooldown;
  file >> organismsCount;

  World w(width, height, startX, startY);

  w.setDay(day-1);
  w.setId(wId);

  int i;
  for (i = 0; i < organismsCount; i++) {
    std::string type;
    int id;
    int strength;
    int initiative;
    int x;
    int y;
    int parent1Id;
    int parent2Id;
    int age;
    int previousX;
    int previousY;
    bool born;
    bool alive;

    file >> type;
    file >> id;
    file >> strength;
    file >> initiative;
    file >> x;
    file >> y;
    file >> parent1Id;
    file >> parent2Id;
    file >> age;
    file >> previousX;
    file >> previousY;
    file >> born;
    file >> alive;

    Organism *organism = nullptr;
    if (type == "Sheep") {
      organism = new Sheep(&w, x, y, parent1Id, parent2Id);
    } else if (type == "Antelope") {
      organism = new Antelope(&w, x, y, parent1Id, parent2Id);
    } else if (type == "SosnowskyHogweed") {
      organism = new SosnowskyHogweed(&w, x, y);
    } else if (type == "Guarana") {
      organism = new Guarana(&w, x, y);
    } else if (type == "Grass") {
      organism = new Grass(&w, x, y);
    } else if (type == "Wolf") {
      organism = new Wolf(&w, x, y, parent1Id, parent2Id);
    } else if (type == "Fox") {
      organism = new Fox(&w, x, y, parent1Id, parent2Id);
    } else if (type == "Turtle") {
      organism = new Turtle(&w, x, y, parent1Id, parent2Id);
    } else if (type == "Human") {
      organism = new Human(&w, x, y, parent1Id, parent2Id);
    } else if (type == "NightshadeBerries") {
      organism = new NightshadeBerries(&w, x, y);
    } else if (type == "Milkweed") {
      organism = new Milkweed(&w, x, y);
    }

    organism->setStrength(strength);
    organism->setInitiative(initiative);
    organism->setId(id);
    organism->setAge(age);
    organism->setPreviousX(previousX);
    organism->setPreviousY(previousY);
    organism->setBorn(born);
    organism->setAlive(alive);

    organisms.push_back(organism);
  }

  file.close();

  w.init(organisms);
  w.setId(wId);

  if (w.getHuman() != nullptr) {
    w.getHuman()->setAbilityCooldown(humanAbilityCooldown);
    w.getHuman()->setAbilityDuration(humanAbilityDuration);
  }

  return w;
}

void Saving::setWorld(World* w) { this->world = w; }

File: src/Saving/Saving.java
package Saving;

import Animal.Animal;
import Human.Human;
import Organism.Organism;
import OrganismManager.*;
import Window.Window;
import World.*;

import javax.swing.*;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;

public class Saving {

    private World world;
    private final Window window;
    private final JFrame frame;

    private final String fileExtension = ".txt";

    public Saving(Window window, JFrame frame) {
        this.window = window;
        this.frame = frame;
    }

    public void setWorld(World world) {
        this.world = world;
    }

    public void save() {
        String fileName = this.getFileName();

        PrintWriter writer = null;
        try {
            writer = new PrintWriter(fileName, "UTF-8");
        } catch (FileNotFoundException | UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }

        writer.println(world.getWorldId());
        writer.println(world.getWorldSettings().worldType());
        writer.println(world.getDay());
        writer.println(world.getWorldSettings().width());
        writer.println(world.getWorldSettings().height());

        Human human = world.getHuman();
        if (human != null) {
            writer.println(human.getAbilityDuration());
            writer.println(human.getAbilityCooldown());
        } else {
            writer.println(0);
            writer.println(0);
        }

        ArrayList<Organism> organisms = world.getOrganisms();
        writer.println(organisms.size());
        for (Organism organism : organisms) {
            writer.print(organism.getType() + " ");
            writer.print(organism.getData().getId() + " ");
            writer.print(organism.getData().getStrength() + " ");
            writer.print(organism.getCurrentField().getNumber() + " ");
            writer.print(organism.getData().getParent1Id() + " ");
            writer.print(organism.getData().getParent2Id() + " ");
            writer.print(organism.getData().getAge() + " ");

            if (organism instanceof Animal) {
                writer.print(((Animal) organism).getLastField().getNumber() + " ");
            } else {
                writer.print("-1 ");
            }

            writer.print(organism.getData().isBorn() + " ");
            writer.println(organism.getData().isAlive());
        }

        writer.close();

        this.showSavedDialog();
    }

    public World load(String fileName) {
        fileName += this.fileExtension;

        Path path = Paths.get(fileName);
        String[] read;
        try {
            read = Files.readAllLines(path).toArray(new String[0]);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        for (String line : read) {
            System.out.println(line);
        }

        String id = read[0];
        String worldType = read[1];
        int day = Integer.parseInt(read[2]);
        int width = Integer.parseInt(read[3]);
        int height = Integer.parseInt(read[4]);
        int abilityDuration = Integer.parseInt(read[5]);
        int abilityCooldown = Integer.parseInt(read[6]);

        WorldSettings worldSettings = new WorldSettings(WorldType.valueOf(worldType), width, height);
        if (worldType.equals("HEXAGONAL")) {
            this.world = new HexWorld(window, worldSettings);
        } else if (worldType.equals("RECTANGULAR")) {
            this.world = new RectangleWorld(window, worldSettings);
        }

        this.world.placeFields();
        this.world.setWorldId(id);
        this.world.setDay(day);

        ArrayList<Organism> organisms = new ArrayList<Organism>();
        int organismsCount = Integer.parseInt(read[7]);

        for (int i = 0; i < organismsCount; i++) {
            String[] organismData = read[8 + i].split(" ");
            String type = organismData[0];
            int organismId = Integer.parseInt(organismData[1]);
            int strength = Integer.parseInt(organismData[2]);
            int fieldNumber = Integer.parseInt(organismData[3]);
            int parent1Id = Integer.parseInt(organismData[4]);
            int parent2Id = Integer.parseInt(organismData[5]);
            int age = Integer.parseInt(organismData[6]);
            int lastFieldNumber = Integer.parseInt(organismData[7]);
            boolean born = Boolean.parseBoolean(organismData[8]);
            boolean alive = Boolean.parseBoolean(organismData[9]);

            Organism organism = null;
            if (type.equals("Human")) {
                organism = new Human(this.world, this.world.getField(fieldNumber), parent1Id, parent2Id);
                organism.getData().setId(organismId);
                organism.getData().setStrength(strength);
                organism.getData().setAge(age);
                organism.getData().setBorn(born);
                organism.getData().setAlive(alive);
                ((Human) organism).setAbilityDuration(abilityDuration);
                ((Human) organism).setAbilityCooldown(abilityCooldown);
            } else {
                organism = OrganismManager.createOrganism(OrganismType.valueOf(type), this.world, this.world.getField(fieldNumber));
                organism.getData().setId(organismId);
                organism.getData().setStrength(strength);
                organism.getData().setParent1Id(parent1Id);
                organism.getData().setParent2Id(parent2Id);
                organism.getData().setAge(age);
                organism.getData().setBorn(born);
                organism.getData().setAlive(alive);

                if (organism instanceof Animal) {
                    ((Animal) organism).setLastField(this.world.getField(lastFieldNumber));
                }
            }
            organisms.add(organism);
        }

        // get maximum organism id
        int maxId = 0;
        for (Organism organism : organisms) {
            if (organism.getData().getId() > maxId) {
                maxId = organism.getData().getId();
            }
        }
        this.world.setNextOrganismId(maxId+1);
        this.world.addOrganisms(organisms);

        this.window.setDayLabel(this.world.getDay());
        this.window.setHumanAbilityDurationLabel(abilityDuration+1);
        if (abilityDuration == 0) {
            this.window.setHumanAbilityCooldownLabel(abilityCooldown);
        }

        return this.world;
    }

    private void showSavedDialog() {
        JOptionPane.showMessageDialog(this.frame, "Simulation saved to: " + this.getFileName(), "Saved", JOptionPane.INFORMATION_MESSAGE);
    }

    private String getFileName() {
        return this.world.getWorldId() + "-" + this.world.getDay() + this.fileExtension;
    }
}
